daliuge-engine/dlg/manager/proc_daemon.py

```diff
@@ -245,14 +260,40 @@ def nm_callback(zeroconf, service_type, name, state_change):
                     # we avoid hanging in here if the MM is down already but
                     # we are trying to remove our NM who has just disappeared
                     if not self._shutting_down:
-                        try:
-                            mm_client.remove_node(server)
-                        finally:
-                            del node_managers[name]
+                        nm_assigner.remove_nm(name)
+
+            def dim_callback(zeroconf, service_type, name, state_change):
```

Comment:
`dim_callback` is sufficiently similar to `nm_callback` that they shouldn't be duplicated. Let's add a more generic function that can then be specialised for nm and dim callback-ing. Something like:

```python
def _callback(zeroconf, service_type, name, state_change, adder, remover, accessor):
   ....

nm_callback = functools.partial(_callback,  adder=nm_assigner.add_nm, remover=nm_assigner.remove_nm, accessor=nm_assigner.NMs)
dim_callback = functools.partial(_callback,  adder=nm_assigner.add_dim, remover=nm_assigner.remove_dim, accessor=nm_assigner.DIMs)
```